[breaking] Rewrite using atomic-fs-blob-store

gmaclennan opened the pull request:
This uses [atomic-fs-blob-store](https://github.com/binded/atomic-fs-blob-store) which in turn uses [fs-write-stream-atomic](https://github.com/npm/fs-write-stream-atomic) which has way more tests for broken writes, especially edge cases on windows.

I wrote more tests, which fail on the `flush` branch. This rewrite simplifies things a lot I think and should be more robust. I needed to use a fork of fs-write-stream-atomic to allow for custom names for tmp files, so that the `list()` function will work.

I also removed the `_` prefix from the `list()` function, since this is now a tested part of the API that we depend upon and should not be considered an internal function.

There are two failing tests because of a [breaking change](https://github.com/maxogden/abstract-blob-store/issues/15) that was introduced in the latest abstract-blob-store without a major version bump (not found file errors do not have a `notFound=true` property set)

hackergrrl committed:
fix: [BREAKING] don't emit "end" until staging write is complete

hackergrrl committed:
fix: maintain original write stream callback behaviour

hackergrrl committed:
fix: report non-exists error as per abstract-blob-store test suite

gmaclennan committed:
Rewrite to use atomic-fs-blob-store

gmaclennan committed:
Tests

gmaclennan committed:
Fix list test

gmaclennan committed:
Pin abstract-blob-store to 3.2.0 to fix not found test

abstract-blob-store 3.3.0 introduced a breaking change, requiring not found errors to have a notFound property set
https://github.com/maxogden/abstract-blob-store/pull/31

gmaclennan committed:
Fix readme example

gmaclennan committed:
Add test for readme example just to make sure

gmaclennan committed:
Add list() function to readme

gmaclennan committed:
Add changelog

gmaclennan committed:
Fix changelog compare

okdistribute commented on package.json:
```diff
@@ -15,12 +15,14 @@
   "keywords": [],
   "dependencies": {
-    "fs-blob-store": "^5.2.1",
+    "@digidem/atomic-fs-blob-store": "^5.3.0",
```
do we need to use a fork of this or can we use the npm published version that isn't digidem?

gmaclennan replied:
We are Using that fork in order to use a fork of fs-create-write-stream-atomic which adds a new option to allow a custom naming function for the temp files. The opts are passed through by atomic-FS-blob-Store so if I get a PR accepted as a minor release it would just work. 
We need the custom naming function so that the list function works (so it doesn’t list temp files left by failed rights).

okdistribute replied:
sioundssounds good

okdistribute commented on index.js:
```diff
@@ -1,96 +1,98 @@
 
-// TODO: opts to choose whether to use staging area
-BlobStore.prototype.createWriteStream = function (opts, done) {
+BlobStore.prototype.list = function (cb) {
```
._list with an underscore is depended upon by blob-store-replication stream so we'd need to update that code too https://github.com/noffle/blob-store-replication-stream/blob/master/index.js#L132

gmaclennan replied:
Yeah happy to leave this as a non-breaking change and keep as _write.

okdistribute replied:
great, looks good to me then otherwise

hackergrrl replied:
Leave it as `list` but also add `BlobStore.prototype._list = BlobStore.prototype.list`, and we can patch the replication stream module at our leisure.

karissa commented:
@noffle if it looks okay to you can we merge and then change list to _list before publishing? this is on gregor's branch but I don't know if he'll have time to look at this since he's in the field

hackergrrl commented on README.md:
```diff
@@ -36,6 +39,14 @@ ws.end(function() {
+
+See https://github.com/maxogden/abstract-blob-store and in addition:
+
+### store.list(cb)
```
You'll need to patch `blob-store-replication-stream` if you want this to work.

I suggest the patch checks for both `list` and `_list` and calls whatever it finds (`list` preferred), for backwards compatibility.

hackergrrl commented on package.json:
```diff
@@ -15,12 +15,14 @@
   "keywords": [],
   "dependencies": {
-    "fs-blob-store": "^5.2.1",
+    "@digidem/atomic-fs-blob-store": "^5.3.0",
```
Bonus points if you send a PR to the original repo!

hackergrrl commented on index.js:
```diff
@@ -1,96 +1,98 @@
-  var store = this._getStore(subdir)
-  return store.exists(opts, done)
+BlobStore.prototype.exists = function (opts, cb) {
+  if (typeof opts === 'string') opts = {key: opts}
```
`if (typeof cb !== 'function') cb = noop` (or fail if not given)

gmaclennan replied:
This is handled by atomic-fs-blob-store

hackergrrl commented on index.js:
```diff
@@ -1,96 +1,98 @@
-  var store = this._getStore(subdir)
-  return store.remove(opts, done)
+BlobStore.prototype.remove = function (opts, cb) {
+  if (typeof opts === 'string') opts = {key: opts}
```
`if (typeof cb !== 'function') cb = noop`

gmaclennan replied:
This is passed through to atomic-fs-blob-store and handled there in accordance to abstract-blob-store specs

hackergrrl commented on index.js:
```diff
@@ -1,96 +1,98 @@
+BlobStore.prototype._insertSubDirPrefix = function (key) {
+  var prefixLen = this.subDirPrefixLen
+  var parsed = path.parse(key)
+  if (parsed.name.length <= prefixLen) return key
```
The old code used as much of the string as possible as a prefix. I think that way will give better results than this: not doing any prefixing at all on short keys.

hackergrrl replied:
I think you can just remove this one line and the `slice` below will work fine on short keys.

hackergrrl replied:
This will also simplify the removal code below, since you know you can always just pop the top dir off.

gmaclennan replied:
I think there are edge-cases that could creep in (dot-files etc). I changed the way this is done now. There are maybe other ways of doing this but the current way I think I've thought through a few different edge cases.

hackergrrl replied:
What does the subdir suffix add to this?

Can you say more about these edge cases you're thinking of? How do dotfiles cause problems?

gmaclennan replied:
I wrote a breaking test before this commit. The issue is a naming collision between two file named `fo` and `foobar`. Internally these collide (filename with same name as folder). 
The other issue if you are just using the prefix is two files with keys `foobar.txt` and `fo/obar.txt` which would overwrite each other. (This is with a súbdito prefix length of 2)

gmaclennan replied:
I wrote a breaking test before this commit. The issue is a naming collision between two file named `fo` and `foobar`. Internally these collide (filename with same name as folder). 
The other issue if you are just using the prefix is two files with keys `foobar.txt` and `fo/obar.txt` which would overwrite each other. (This is with a súbdito prefix length of 2)

okdistribute commented on index.js:
```diff
@@ -1,96 +1,98 @@
+  if (parsed.name.length <= prefixLen) return key
+  var dirs = parsed.dir.split('/')
+  if (parsed.name.slice(0, prefixLen) !== dirs.pop()) return key
+  return path.join(dirs.join('/'), parsed.base)
```
`path.sep` here for windows

gmaclennan committed:
Use path.sep not '/'

gmaclennan committed:
Catch error

gmaclennan committed:
Breaking test with funky filename

gmaclennan committed:
Use only first two chars for subdir (like git)

With random base64 filenames, using two chars is 64^2 = 4096 files. Using 3 chars would be 262,144 possibilities, which goes over the FAT32 limit.

gmaclennan committed:
Update breaking test for 2-char subdir

gmaclennan committed:
Use stable identifier for subdirs to fix potential naming conflicts between files and dirs

This also makes the store continue to work even if you change `subDirPrefixLen`

gmaclennan committed:
newline cleanup

gmaclennan committed:
Alias _list to list

noffle commented:
Closing (merging #4 instead, which has some additional patches).
